python/pyspark/sql/udaf.py

```diff
@@ -0,0 +1,897 @@
+    udaf_col_expr = udaf_col._udaf_col  # type: ignore[attr-defined]
+
+    # Get grouping columns
+    grouping_cols = _extract_grouping_columns_from_jvm(jgd)
+
+    # Use Catalyst optimizer path (via Scala pythonAggregatorUDAF)
+    return _apply_udaf_via_catalyst(
+        df,
+        jgd,
+        udaf_func,
+        udaf_col_expr,
+        grouping_cols if grouping_cols else None,
+    )
+
+
+def _get_max_key_for_random_grouping(df: "DataFrame") -> int:
+    """Get max key for random grouping based on Spark config or partition count."""
+    try:
+        spark_session = df.sparkSession
+        shuffle_partitions = int(spark_session.conf.get("spark.sql.shuffle.partitions", "200"))
+        num_partitions = df.rdd.getNumPartitions()
+        return max(shuffle_partitions, num_partitions, 1)
+    except Exception:
+        return 200
+
+
+def _convert_results_to_arrow(results: List[Any], return_type: DataType) -> Any:
+    """Convert a list of result values to Arrow array based on return type."""
+    import pyarrow as pa
+    from pyspark.sql.pandas.types import to_arrow_type
+    from pyspark.sql.conversion import LocalDataToArrowConversion
+
+    # Use existing conversion utilities for accurate type handling
+    arrow_type = to_arrow_type(return_type)
+    converter = LocalDataToArrowConversion._create_converter(return_type, nullable=True)
+
+    if converter is not None:
+        converted_results = [converter(r) for r in results]
+    else:
+        converted_results = results
+
+    return pa.array(converted_results, type=arrow_type)
+
+
+def _create_reduce_func(
+    serialized_aggregator: bytes,
+    max_key: int,
+    num_grouping_cols: int,
+):
+    """Create reduce function for mapInArrow."""
+
+    def reduce_func(iterator):
+        import pyarrow as pa
+        import cloudpickle
+        import random
+
+        agg = cloudpickle.loads(serialized_aggregator)
+        group_buffers = {}
+        value_col_idx = num_grouping_cols
+
+        for batch in iterator:
+            if (
+                isinstance(batch, pa.RecordBatch)
+                and batch.num_columns > value_col_idx
+                and batch.num_rows > 0
+            ):
+                value_col = batch.column(value_col_idx)
+
+                for row_idx in range(batch.num_rows):
+                    # Extract grouping key (None for non-grouped case, tuple for grouped case)
+                    grouping_key = (
+                        tuple([batch.column(i)[row_idx].as_py() for i in range(num_grouping_cols)])
+                        if num_grouping_cols > 0
+                        else None
+                    )
+
+                    value = value_col[row_idx].as_py()
+
+                    if grouping_key not in group_buffers:
+                        group_buffers[grouping_key] = agg.zero()
+
+                    if value is not None:
+                        group_buffers[grouping_key] = agg.reduce(group_buffers[grouping_key], value)
```

Comment:
`group_buffers` buffers all the aggregators within a partition, it will cause memory issue if the cardinality is large.
A reasonable physical plan should sort the partition by the key, and then output the partial aggregation result after finishing each group

Reply:
thanks for the suggestion. I will take a look on different aggregateExec implementations.